src/main/java/com/lpvs/service/LPVSLicenseService.java

```diff
@@ -212,10 +232,87 @@ public LPVSLicense findLicenseByName(String name) {
+     * @param licenseSpdxId The license SPDX ID to search for
+     * @param licenseName The license name to use if the license is not found in the database
+     * @return The license object that was found or created
+     */
+    public LPVSLicense getLicenseBySpdxIdAndName(
+            String licenseSpdxId, Optional<String> licenseName) {
+        String licName = licenseName.orElse(licenseSpdxId);
+        // check by license SPDX ID
+        LPVSLicense lic = findLicenseBySPDX(licenseSpdxId);
+        if (null == lic) {
+            // check by license name and alternative names
+            lic = findLicenseByName(licName);
+        }
+        // create new license
+        if (lic == null) {
+
+            if (osoriDbUrl != null && !osoriDbUrl.trim().isEmpty()) {
+                // check OSORI DB and try to find the license there
+                try {
+                    HttpURLConnection connection =
+                            osoriConnection.createConnection(osoriDbUrl, licenseSpdxId);
+                    connection.setRequestMethod("GET");
+                    connection.connect();
+
+                    if (connection.getResponseCode() != 200) {
+                        throw new Exception(
+                                "HTTP error code ("
+                                        + connection.getResponseCode()
+                                        + "): "
+                                        + connection.getResponseMessage());
+                    }
+
+                    BufferedReader in =
+                            LPVSPayloadUtil.createBufferReader(
+                                    LPVSPayloadUtil.createInputStreamReader(
+                                            connection.getInputStream()));
+                    String inputLine;
+                    StringBuffer response = new StringBuffer();
+
+                    while ((inputLine = in.readLine()) != null) {
+                        response.append(inputLine);
+                    }
+                    in.close();
+
+                    // if found, create new license with field values taken from OSORI DB
+                    lic = LPVSPayloadUtil.convertOsoriDbResponseToLicense(response.toString());
+                } catch (Exception e) {
+                    log.error("Error connecting OSORI DB: " + e.getMessage());
+                }
+            }
+
+            if (lic == null) {
+                // if not found, create new license with default field values
+                lic = new LPVSLicense();
+                lic.setSpdxId(licenseSpdxId);
+                lic.setLicenseName(licName);
+                lic.setAlternativeNames(null);
+                lic.setAccess("UNREVIEWED");
+            }
+            // save new license and add it to the license list
+            lic = lpvsLicenseRepository.saveAndFlush(lic);
+        }
+        addLicenseToList(lic);
+        return lic;
+    }
```

Comment:
What about refactor the code in terms of the improvement points below? 

- Improvement Points:

1. The code is quite long and can be simplified by breaking it down into smaller methods.
2. The code is checking for the license in multiple places, which can be refactored into a single method.

Please refer to the example code below and check if it is possible. 😁
```java
public LPVSLicense getLicenseBySpdxIdAndName(String licenseSpdxId, Optional<String> licenseName) {
    String licName = licenseName.orElse(licenseSpdxId);

    // Check if the license exists in the database
    LPVSLicense lic = findLicense(licenseSpdxId, licName);

    // If not found, check OSORI DB and create a new license if found
    if (lic == null) {
        lic = findLicenseInOsoriDb(licenseSpdxId);
    }

    // If still not found, create a new license with default field values
    if (lic == null) {
        lic = createNewLicense(licenseSpdxId, licName);
    }

    // Save the license and add it to the license list
    lic = lpvsLicenseRepository.saveAndFlush(lic);

    return lic;
}

private LPVSLicense findLicense(String licenseSpdxId, String licName) {
    LPVSLicense lic = findLicenseBySPDX(licenseSpdxId);
    if (lic == null) {
        lic = findLicenseByName(licName);
    }
    return lic;
}

private LPVSLicense findLicenseInOsoriDb(String licenseSpdxId) {
    if (osoriDbUrl == null || osoriDbUrl.trim().isEmpty()) {
        return null;
    }

    try {
        HttpURLConnection connection = osoriConnection.createConnection(osoriDbUrl, licenseSpdxId);
        connection.setRequestMethod("GET");
        connection.connect();

        if (connection.getResponseCode()!= 200) {
            throw new Exception("HTTP error code (" + connection.getResponseCode() + "): " + connection.getResponseMessage());
        }

        String response = LPVSPayloadUtil.convertInputStreamToString(connection.getInputStream());
        return LPVSPayloadUtil.convertOsoriDbResponseToLicense(response);
    } catch (Exception e) {
        log.error("Error connecting OSORI DB: " + e.getMessage());
        return null;
    }
}

private LPVSLicense createNewLicense(String licenseSpdxId, String licName) {
    LPVSLicense lic = new LPVSLicense();
    lic.setSpdxId(licenseSpdxId);
    lic.setLicenseName(licName);
    lic.setAlternativeNames(null);
    lic.setAccess("UNREVIEWED");
    return lic;
}
```

Reply:
@tiokim  Thank you for the suggestion. I will apply the proposed changes in the current PR.

Reply:
Thank you for modifying it according to the review comments. I think the test code is also well written. One question is, is there a reason why the functions were public but not private?

Reply:
I make these functions public for providing better testability and cover them with unit tests (to avoid the usage of reflection to get a reference to the private methods). If you have other opinion, please let me know.